components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentEvaluatorTest.kt

```diff
@@ -642,4 +502,140 @@ class ExperimentEvaluatorTest {
+    }
+
+    @Test(expected = AssertionError::class)
+    fun `test selectBranchByWeight() asserts lower out-of-bounds`() {
+        val branches = listOf(
+            Experiment.Branch("1", 2)
+        )
+        val evaluator = ExperimentEvaluator()
+        evaluator.selectBranchByWeight(2, branches)
+    }
+
+    @Test
+    fun `test pickActiveBranch()`() {
+        val e = ExperimentEvaluator()
+
+        var experiment = testDefaultExperiment(
+            branches = listOf(
+                Experiment.Branch("1", 2)
+            )
+        )
+
+        assertEquals("1", e.pickActiveBranch(experiment, { _, _ -> 0 }).name)
+        assertEquals("1", e.pickActiveBranch(experiment, { _, _ -> 1 }).name)
+
+        experiment = testDefaultExperiment(
+            branches = listOf(
+                Experiment.Branch("1", 1),
+                Experiment.Branch("2", 2)
+            )
+        )
+
+        assertEquals("1", e.pickActiveBranch(experiment, { _, _ -> 0 }).name)
+        assertEquals("2", e.pickActiveBranch(experiment, { _, _ -> 1 }).name)
+        assertEquals("2", e.pickActiveBranch(experiment, { _, _ -> 2 }).name)
+    }
+
+    private fun evaluate(
+        diceValue: Int,
+        descriptor: ExperimentDescriptor,
+        experiments: List<Experiment>
+    ): ActiveExperiment? {
+        val e = ExperimentEvaluator(object : ValuesProvider() {
+            override fun getRandomBranchValue(lower: Int, upper: Int): Int {
+                return diceValue
+            }
+            override fun getClientId(context: Context): String = "c641eacf-c30c-4171-b403-f077724e848a"
+        })
+        return e.evaluate(context, descriptor, experiments)
+    }
+
+    @Test
+    fun `test evaluating branch choices`() {
+        testReset()
+
+        var experiments = listOf(
+            testDefaultExperiment(
+                id = "id",
+                branches = listOf(
+                    Experiment.Branch("1", 2)
+                )
+            )
+        )
+        val descriptor = ExperimentDescriptor("id")
+        assertEquals(
+            ActiveExperiment(experiments[0], "1"),
+            evaluate(0, descriptor, experiments)!!
+        )
+        assertEquals(
+            ActiveExperiment(experiments[0], "1"),
+            evaluate(0, descriptor, experiments)!!
+        )
+
+        experiments = listOf(
+            testDefaultExperiment(
+                id = "other"
+            ),
+            testDefaultExperiment(
+                id = "id",
+                branches = listOf(
+                    Experiment.Branch("1", 1),
+                    Experiment.Branch("2", 1)
+                )
+            )
+        )
+        assertEquals(
+            ActiveExperiment(experiments[1], "1"),
+            evaluate(0, descriptor, experiments)!!
+        )
+        assertEquals(
+            ActiveExperiment(experiments[1], "2"),
+            evaluate(1, descriptor, experiments)!!
+        )
+    }
 }
```

Comment:
nit: no newline at end of file (the linter doesn't require this but github seems to notice and most of our files have a blank line at the end)